codec-http/src/main/java/io/netty/handler/codec/http/HttpRequestCompressor.java

```diff
@@ -0,0 +1,361 @@
+     * @param preferredEncoding the preferred encoding.
+     * if unavailable, the default encoding {@link #DEFAULT_ENCODING} will be used.
+     * @param contentSizeThreshold the size in byte the http body must have before compressing the request
+     * @see #HttpRequestCompressor(java.lang.String, int, io.netty.handler.codec.compression.CompressionOptions...)
+     */
+    public HttpRequestCompressor(String preferredEncoding, int contentSizeThreshold) {
+        this(preferredEncoding, contentSizeThreshold, (CompressionOptions[]) null);
+    }
+
+    /**
+     * new instance using the preferred encoding, threshold and compression options.
+     * @param preferredEncoding the preferred encoding.
+     * if unavailable, the default encoding {@link #DEFAULT_ENCODING} will be used.
+     * @param contentSizeThreshold the size in byte the http body must have before compressing the request
+     * @param compressionOptions the desired compression options to use.
+     * if {@code null} or empty, the defaults will be used.
+     * @see StandardCompressionOptions#brotli() default brotli options
+     * @see StandardCompressionOptions#gzip() default gzip options
+     * @see StandardCompressionOptions#zstd() default zstd options
+     * @see StandardCompressionOptions#deflate() default deflate options
+     */
+    public HttpRequestCompressor(String preferredEncoding, int contentSizeThreshold,
+            CompressionOptions... compressionOptions) {
+        ObjectUtil.checkNonEmpty(preferredEncoding, "preferredEncoding");
+        this.contentSizeThreshold = ObjectUtil.
+                checkPositiveOrZero(contentSizeThreshold, "contentSizeThreshold");
+
+        if (!SUPPORTED_ENCODINGS.contains(preferredEncoding)) {
+            throw new IllegalArgumentException(
+                    String.format("Unsupported encoding %s. Supported encodings are: %s",
+                            preferredEncoding, StringUtil.join(",", SUPPORTED_ENCODINGS)));
+        }
+
+        GzipOptions gzipOptions = null;
+        DeflateOptions deflateOptions = null;
+        BrotliOptions brotliOptions = null;
+        ZstdOptions zstdOptions = null;
+        for (CompressionOptions compressionOption : Optional.ofNullable(compressionOptions)
+                .orElseGet(() -> new CompressionOptions[0])) {
+            if (compressionOption instanceof BrotliOptions) {
+                brotliOptions = (BrotliOptions) compressionOption;
+            } else if (compressionOption instanceof GzipOptions) {
+                gzipOptions = (GzipOptions) compressionOption;
+            } else if (compressionOption instanceof DeflateOptions) {
+                deflateOptions = (DeflateOptions) compressionOption;
+            } else if (compressionOption instanceof ZstdOptions) {
+                zstdOptions = (ZstdOptions) compressionOption;
+            } else {
+                log.info("ignoring unsupported compression option {}",
```

Comment:
The compression option would be "unrecognized" in this case. I think we should move this to after we've determined that the desired encoding is available, and then ensure that all the options are of the relevant type.

Reply:
If an exception is to be thrown when the preferred encoding is not available, an array of options is not needed anyway. Also, you cannot pass an instance of `BrotliOptions` if Brotli is not available, because an exception would have already been thrown if the user tried to create an instance of `BrotliOptions`. So it is fine to first determine the concrete type of the given `CompressionOptions`. I can also check if the encoding and the given `CompressionOptions` match (e.g. "gzip" and `GzipOptions` are okay but not "gzip" and `DeflateOptions`).